scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/InMemoryConnectionStore.java

```diff
@@ -147,105 +149,148 @@ public void onEviction(Connection staleConnection) {
+							remove(conFromLocalCache.getPeerAddress());
+							// TODO: should we send a fatal alert to peer in this case?
+						}
+	
+						return null;
+	
+					} else if (conFromLocalCache == null) {
+						// this probably means that we are taking over the session from a failed node
+							return new Connection(ticket);
+							// connection will be put to first level cache as part of
+							// the abbreviated handshake
+					} else {
+						// resume connection found in local cache (i.e. this store)
+						return conFromLocalCache;
 					}
-
-					return null;
-
-				} else if (conFromLocalCache == null) {
-					// this probably means that we are taking over the session from a failed node
-						return new Connection(ticket);
-						// connection will be put to first level cache as part of
-						// the abbreviated handshake
-				} else {
-					// resume connection found in local cache (i.e. this store)
-					return conFromLocalCache;
+				} finally {
+					conLock.writeLock().unlock();
 				}
 			}
 		}
 	}
 
-	private synchronized Connection findLocally(final SessionId id) {
-
-		return connections.find(new Predicate<Connection>() {
-			@Override
-			public boolean accept(final Connection connection) {
-				DTLSSession session = connection.getEstablishedSession();
-				return session != null && id.equals(session.getSessionIdentifier());
-			}
-		});
+	private Connection findLocally(final SessionId id) {
+		try {
+			conLock.readLock().lock();
+			return connections.find(new Predicate<Connection>() {
+				@Override
+				public boolean accept(final Connection connection) {
+					DTLSSession session = connection.getEstablishedSession();
+					return session != null && id.equals(session.getSessionIdentifier());
+				}
+			});
+		} finally {
+			conLock.readLock().unlock();
+		}
 	}
 
 	@Override
-	public synchronized void markAllAsResumptionRequired() {
-		for (Iterator<Connection> iterator = connections.values(); iterator.hasNext(); ) {
-			Connection c = iterator.next();
-			if (c != null){
-				c.setResumptionRequired(true);
+	public void markAllAsResumptionRequired() {
+		try {
+			conLock.writeLock().lock();
+			for (Iterator<Connection> iterator = connections.values(); iterator.hasNext(); ) {
+				Connection c = iterator.next();
+				if (c != null){
+					c.setResumptionRequired(true);
+				}
 			}
+		} finally {
+			conLock.writeLock().unlock();
 		}
 	}
 
 	@Override
-	public synchronized int remainingCapacity() {
-		return connections.remainingCapacity();
+	public int remainingCapacity() {
+		try {
+			conLock.readLock().lock();
+			return connections.remainingCapacity();
+		} finally {
+			conLock.readLock().unlock();
+		}
+		
 	}
 
 	@Override
-	public synchronized Connection get(final InetSocketAddress peerAddress) {
-		return connections.get(peerAddress);
+	public Connection get(final InetSocketAddress peerAddress) {
+		try {
+			conLock.readLock().lock();
+			return connections.get(peerAddress);
```

Comment:
Unfortunately, `LeastRecentlyUsedCache.get(K key)` seems not to be clearly "read only".

```
		if (key == null) {
			return null;
		}
		CacheEntry<K, V> entry = cache.get(key);
		if (entry == null) {
			return null;
		} else if (entry.isStale(expirationThreshold)) {
			cache.remove(entry.getKey());
			entry.remove();
			return null;
		} else {
			entry.recordAccess(header);
			return entry.getValue();
		}

```